### PR TITLE
ADS Control Flow: Move shouldRespond & registerEvent to receive() goroutine

### PR DESCRIPTION
**Pull request by adiprerepa**
Moving `shouldRespond` to the `receive` goroutine.

### CODE REVIEW

File: pilot/pkg/xds/rds.go [hunk 1]
```diff
@@ -58,6 +58,8 @@ func rdsNeedsPush(req *model.PushRequest) bool {
 }
 
 func (c RdsGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, req *model.PushRequest) model.Resources {
+	proxy.Lock()
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/rds.go:61]: why do we need this lock

Comment (adiprerepa) [reply pilot/pkg/xds/rds.go:61]: race without it

Comment (hzxuzhonghu) [reply pilot/pkg/xds/rds.go:61]: Then i doubt the meaning of move it to the receive go routine

**Pull request closed without merging**